### PR TITLE
Remove workflow-based scene transition design

### PR DESCRIPTION
**Pull request by Jon-edge**
The login UI design uses workflows, which are composed of one or more
scenes that are indexed by a currentSceneIndex. The view transitions
from scene to scene by incrementing the currentSceneIndex with
ReduxAction WORKFLOW_NEXT. This design of blindly transitioning to the
next scene in the workflow allows for scenarios in which scenes may be
skipped when multiple WORKFLOW_NEXT actions are made. Most workflows
consist only of a single scene.

Replacing the workflow-based grouped scene transition design with a
flat hierarchy of scenes which are explicitly called during each
transition guards against the possibility of unintentionally
transitioning past the intended scene.

### CODE REVIEW

File: packages/edge-login-ui-rn/src/constants/Scenes.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+export interface Scene {
+  title: string
+  subTitle: string
+}
+
+export enum SceneNames {
```
Comment (swansontec) [inline packages/edge-login-ui-rn/src/constants/Scenes.ts:6]: While it's not "deprecated" per-se, nobody really likes the TypeScript `enum` type, because it's not part of the ECMA-262 JavaScript standard. Generally speaking, TypeScript is pretty much "plain JavaScript plus types", so you can just strip the types to get valid JS. The `enum` keyword, on the other hand, actually requires code-generation, since it's a new language feature that exists at run-time.

Instead of using `enum`, the modern thing is to just use `type SomeEnum = 'a' | 'b' | 'c'`, which gives the same benefits while being a pure compile-time construct. This is because, if you have `let x: SomeEnum`, you can safely write code like `x = 'a'`. If you make a typo like `x = 'd'` or `if (x === 'd')`, TypeScript will complain because the types say that can't happen. So, the raw strings are actually safe.

Now, I realize that in most other languages raw strings are incredibly dangerous. Things like `x = 'a'` are a huge red flag, because there is no type-checking. Therefore, the safe move in C / C++ / C# etc. is to define some variables with the constants, and use those instead. However, this conventional wisdom doesn't apply to TypeScript, where we *do* have type checking on the raw strings! Plus, the extra constants just add size & extra work for the interpreter (unlike a compiled language where they might increase performance).

In other words, the safest & most efficient thing to do is to drop the `enum` keyword and just use raw strings. So, please remove `SceneNames` and just use the strings. It's OK, I promise.

File: packages/edge-login-ui-rn/src/actions/CreateAccountActions.ts [hunk 2]
```diff
@@ -87,6 +87,7 @@ export function validateConfirmPassword(confirmPassword: string) {
   }
 }
 export function validatePassword(data: string) {
+  console.log('validatePassword')
```
Comment (swansontec) [inline packages/edge-login-ui-rn/src/actions/CreateAccountActions.ts:90]: Please remove the extra logging.

File: packages/edge-login-ui-rn/src/actions/CreateAccountActions.ts [hunk 3]
```diff
@@ -131,9 +132,11 @@ export function validatePassword(data: string) {
 }
 
 export function createUser(data: CreateUserData) {
+  console.log('createUser')
```
Comment (swansontec) [inline packages/edge-login-ui-rn/src/actions/CreateAccountActions.ts:135]: Remove logging.

File: packages/edge-login-ui-rn/src/components/common/Header.tsx [hunk 4]
```diff
@@ -4,25 +4,24 @@ import LinearGradient from 'react-native-linear-gradient'
 
 import s from '../../common/locales/strings'
 import T from '../../components/common/FormattedText'
-import { connect } from '../../components/services/ReduxStore'
+// import { connect } from '../../components/services/ReduxStore'
```
Comment (swansontec) [inline packages/edge-login-ui-rn/src/components/common/Header.tsx:7]: Please remove the unused lines, instead of commenting them out.

File: packages/edge-login-ui-rn/src/components/common/Header.tsx [hunk 5]
```diff
@@ -4,25 +4,24 @@ import LinearGradient from 'react-native-linear-gradient'
...
 import { HeaderBackButton } from '../common/HeaderBackButton'
 
 interface OwnProps {
   onBack?: () => void
   onSkip?: () => void
-}
-interface StateProps {
-  subTitle: string
   title: string
+  subTitle?: string
```
Comment (swansontec) [inline packages/edge-login-ui-rn/src/components/common/Header.tsx:20]: Since `subTitle` doesn't even render, let's just remove it completely.

File: packages/edge-login-ui-rn/src/components/navigation/Router.tsx [hunk 6]
```diff
@@ -23,9 +24,9 @@ import { LoadingScreen } from '../screens/LoadingScreen'
 import { NewAccountPasswordScreen } from '../screens/newAccount/NewAccountPasswordScreen'
 import { NewAccountPinScreen } from '../screens/newAccount/NewAccountPinScreen'
 import { NewAccountReviewScreen } from '../screens/newAccount/NewAccountReviewScreen'
+import { NewAccountTosScreen as NewAccountTOSScreen } from '../screens/newAccount/NewAccountTosScreen'
```
Comment (swansontec) [inline packages/edge-login-ui-rn/src/components/navigation/Router.tsx:27]: Let's keep the name as it is, so `NewAccountTosScreen` everywhere instead of renaming it here in the import.

File: packages/edge-login-ui-rn/src/components/screens/RecoveryLoginScreen.tsx [hunk 7]
```diff
@@ -10,7 +10,7 @@ import { Dispatch, RootState } from '../../types/ReduxTypes'
 import { LoginAttempt } from '../../util/loginAttempt'
 import { Button } from '../common/Button'
 import { FormField } from '../common/FormField'
-import { Header } from '../common/Header'
+import { HeaderComponent } from '../common/Header'
```
Comment (swansontec) [inline packages/edge-login-ui-rn/src/components/screens/RecoveryLoginScreen.tsx:13]: Can we rename the component to `Header`, so the change stays local to the file instead of leaking out to all these other files?

File: packages/edge-login-ui-rn/src/constants/Scenes.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+export interface Scene {
```
Comment (swansontec) [inline packages/edge-login-ui-rn/src/constants/Scenes.ts:1]: I don't think the `Scene` type is actually used anywhere?

File: packages/edge-login-ui-rn/src/reducers/WorkflowReducer.ts [hunk 9]
```diff
@@ -1,62 +1,115 @@
...
-const initialState: WorkflowState = {
-  currentKey: 'loadingWF',
-  currentSceneIndex: 0
+const initialState: SceneState = {
+  currentScene: SceneNames.loadingScene
 }
 
-export function workflow(
-  state: WorkflowState = initialState,
+// TODO: remove back/next, update to specific scene
```
Comment (swansontec) [inline packages/edge-login-ui-rn/src/reducers/WorkflowReducer.ts:12]: This TODO is done, right?

File: packages/edge-login-ui-rn/src/reducers/WorkflowReducer.ts [hunk 10]
```diff
@@ -1,62 +1,115 @@
...
+export function scene(
```
Comment (swansontec) [inline packages/edge-login-ui-rn/src/reducers/WorkflowReducer.ts:13]: Since this is now the scene reducer, I think we should rename the file to match. I really like the new name, BTW.

**Review by swansontec — CHANGES_REQUESTED** [review]
Great work! This PR looks really good overall. Most of my comments are just minor stylistic things, but the core change here is really solid. I am impressed.

File: packages/edge-login-ui-rn/src/actions/CreateAccountActions.ts [hunk 11]
```diff
@@ -55,15 +55,14 @@ export function checkUsernameForAvailabilty(data: string) {
 }
 
 function isASCII(str: string) {
-  // eslint-disable-next-line no-control-regex
-  return /^[\x00-\x7F]*$/.test(str)
+  return /^[\x20-\x7E]*$/.test(str)
```
Comment (swansontec) [inline packages/edge-login-ui-rn/src/actions/CreateAccountActions.ts:58]: For future reference, we like to see fixes like this as their own commit, not bundled with other work.

File: packages/edge-login-ui-rn/src/constants/Scenes.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  | 'LandingScene'
+  | 'LoadingScene'
+  | 'OtpScene'
+  | 'OtpRepairScene'
+  | 'PasswordScene'
+  | 'PinScene'
+  | 'RecoveryLoginScene'
+  | 'ResecurePasswordScene'
+  | 'ResecurePinScene'
+  | 'SecurityAlertScene'
```
Comment (swansontec) [inline packages/edge-login-ui-rn/src/constants/Scenes.ts:21]: Since there are no constants in this file, I think it belongs in `types/`, not `constants/`.

File: packages/edge-login-ui-rn/src/reducers/RootReducer.ts [hunk 13]
```diff
@@ -10,15 +10,15 @@ import {
 } from './PasswordRecoveryReducer'
 import { passwordStatus, PasswordStatusState } from './PasswordStatusReducer'
 import { previousUsers, PreviousUsersState } from './PreviousUsersReducer'
-import { workflow, WorkflowState } from './WorkflowReducer'
+import { scene, SceneState } from './WorkflowReducer'
```
Comment (swansontec) [inline packages/edge-login-ui-rn/src/reducers/RootReducer.ts:13]: Since it's `scene` and `SceneState`, I think the file should be renamed to `SceneReducer`. Please do this in its own commit, after the current one. If you rename a file and make changes at the same time, Git sometimes gets confused, so we typically do renames separately.

File: packages/edge-login-ui-rn/src/reducers/WorkflowReducer.ts [hunk 14]
```diff
@@ -1,62 +1,114 @@
...
-        ++currentSceneIndex
-      }
-      return { ...state, currentSceneIndex }
-    }
-
-    // Actions for launching scenes:
     case 'OTP_ERROR':
-      return { ...state, currentKey: 'otpWF', currentSceneIndex: 0 }
+      return {
+        ...state,
```
Comment (swansontec) [inline packages/edge-login-ui-rn/src/reducers/WorkflowReducer.ts:19]: There are a bunch of `...state` entries in this file, but they don't do anything now, because there are no other properties besides `currentScene` that we need to preserve.

**Review by swansontec — CHANGES_REQUESTED** [review]

**Review by swansontec — APPROVED** [review]
